### ISSUE TITLE
Compile with MinGW-64 (G++)

### ISSUE THREAD

**Issue by anonlo-rw** [fpagliughi/sockpp#66]
Hi fpagliughi,

Would you happen to know how to compile a sockpp included CPP file properly with G++? I've tried many including: 
g++ main.cpp -o main -I "sockpp/include/" -L "sockpp/build/Debug/sockpp-static.lib"

Including both the path and the linker file location. The program compiles properly when I have the actual include statement (example: #include <sockpp/tcp_connector.h>) only - but when trying to add a TCP connector (example: sockpp::tcp_connector conn({"localhost", 12345});) - it fails to compile.

**Comment by fpagliughi** [comment]
What platform are you trying this on? From the ".lib" I would assume a flavor of Windows?

I don't know Windows very well, and what limited experience I have with it is all with MSVC. So I'm not sure how much I can help. But...

It appears that you're working out of a build tree without installing it. In that case, the include files are not where you might expect. After install, the tree might look like:
```
usr
└── local
    ├── bin
    ├── include
    │   └── sockpp
    └── lib
```
So there's an `include/sockpp` (not just an `include`)
So on the command line, you tell it to look in `include` ( `-I /usr/local/include` )
And in the source, you specify to look in a `sockpp` subdirectory: ( `#include <sockpp/tcp_connector.h>` )

Then, in Linux, the g++ switch '-L'  tells the linker _where_ to look for files, but the '-l' (lower-case L) tells it the library files. I would assume MinGW does the same?

So, all together, the command line is:

```
$ g++ tcpecho.cpp -o tcpecho -I /my/install/path/include -L /my/install/path/lib -lsockpp
```

**Comment by anonlo-rw** [comment]
Yeah, i'm trying to compile it in windows in order to run on it which I am using as a client where I connect to it from a linux server.

I'm pretty sure it should be very similar due to my experience using the MinGW compiler on both windows and linux being about the same in terms of command line parameters.

I'll give that command a try in a bit, just out at the moment but thank you very much otherwise for your feedback!

**Comment by anonlo-rw** [comment]
So I tried running this compile command:
g++ main.cpp -o main -I "sockpp/include/" -L "sockpp/build/Debug/sockpp-static.lib" -lsockpp

But it was unable to identify what "-lsockpp" is, outputting: cannot find -lsockpp (returned 1 exit status) on Windows.

It is definitely a bit trickier on windows to perform these types of tasks but would you know how I could get MinGW to recognize what -lsockpp is?

**Comment by fpagliughi** [comment]
Sorry I never got back to this. I kept trying to figure out networking on Windows with MinGW, but I couldn't figure out the way different versions of the compiler define different macros. Like some versions seem to define `_WIN32` and others don't?

So I gave up. Windows isn't a huge priority for me, and just getting MSVC working was enough of a chore. If you figured it out, let me know. I'm happy to put up instructions in the README or take whatever PR to make it work.

But, for completeness, with gcc, `-L` typically tells it _where_ to look, so give that a directory. And `-l` tells it what library to link, with just the base file name, no extension. So in your case, together, it should just be:

```
g++ main.cpp -o main -I "sockpp/include/" -L "sockpp/lib" -lsockpp-static
```

**Comment by fpagliughi** [comment]
There were a number of issues with MinGW. The incorrect version of `_WIN32_WINNT` wasn't bringing in the proper set of Winsock functions, and the error return values were incorrect.

The linking should work now; the CMake build is operational and passing all the unit tests. But if you do it manually, you probably need to link in the system Winsock library `ws2_32` also.

There are a couple issues here for MinGW. Let's close this one and move the linking discussion to #68
